sdk/core/System.ClientModel/src/Message/PipelineRequestHeaders.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace System.ClientModel.Primitives;
+
+internal class PipelineRequestHeaders : PipelineMessageHeaders
+{
+    private ArrayBackedPropertyBag<IgnoreCaseString, object> _headers;
+
+    public override bool Remove(string name)
+        => _headers.TryRemove(new IgnoreCaseString(name));
+
+    public override void Set(string name, string value)
+        => _headers.Set(new IgnoreCaseString(name), value);
+
+    public override void Add(string name, string value)
+    {
+        if (_headers.TryAdd(new IgnoreCaseString(name), value, out object? currentValue))
+        {
+            return;
+        }
+
+        switch (currentValue)
+        {
+            case string stringValue:
+                _headers.Set(new IgnoreCaseString(name), new List<string> { stringValue, value });
+                break;
+            case List<string> listValue:
```

Comment:
It seems like we allow adding null values, so don't we need a default branch to update a null value?